core/cache/watcher.go

```diff
@@ -146,6 +148,188 @@ func (w *ConfigWatcher) configChanged(topic string, value interface{}) {
 	w.notify()
 }
 
+type MachineAppLXDProfileWatcher struct {
+	*notifyWatcherBase
+
+	applications map[string]appInfo // unit names for each application
+	machineId    string
+
+	getCharm charmFunc
+}
+
+type charmFunc func(string) (*Charm, error)
+
+type appInfo struct {
+	charmURL     string
+	charmProfile *lxdprofile.Profile
+	units        set.Strings
+}
+
+func newMachineAppLXDProfileWatcher(
+	appTopic, unitTopic, machineId string,
+	applications map[string]appInfo,
+	getCharm charmFunc,
+	hub *pubsub.SimpleHub,
+) *MachineAppLXDProfileWatcher {
+	w := &MachineAppLXDProfileWatcher{
+		notifyWatcherBase: newNotifyWatcherBase(),
+		applications:      applications,
+		getCharm:          getCharm,
+		machineId:         machineId,
+	}
+
+	unsubApp := hub.Subscribe(appTopic, w.applicationCharmURLChange)
+	unsubUnit := hub.Subscribe(unitTopic, w.unitChange)
+	w.tomb.Go(func() error {
+		<-w.tomb.Dying()
+		unsubApp()
+		unsubUnit()
+		return nil
+	})
+
+	return w
+}
+
+// applicationCharmURLChange sends a notification if what is saved for its
+// charm lxdprofile changes.  No notification is sent if the pointer begins
+// and ends as nil.
+func (w *MachineAppLXDProfileWatcher) applicationCharmURLChange(topic string, value interface{}) {
+	w.notifyWatcherBase.mu.Lock()
+	var notify bool
+	defer func(notify *bool) {
+		w.notifyWatcherBase.mu.Unlock()
+		if *notify {
+			w.notify()
+		}
+	}(&notify)
+
+	values, ok := value.([]string)
+	if !ok {
+		logger.Errorf("programming error, value not of type []string")
+		return
+	}
+	if len(values) != 2 {
+		logger.Errorf("programming error, 2 values not provided")
+		return
+	}
+	appName, chURL := values[0], values[1]
+	info, ok := w.applications[appName]
+	if ok {
+		ch, err := w.getCharm(chURL)
+		if err != nil {
+
```

Comment:
What should happen here? I guess return with the error?

Reply:
resolved in pr 9966 lxdprofilewatcher.go